independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java

```diff
@@ -0,0 +1,241 @@
+package io.quarkus.bootstrap.classloading;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.function.BiConsumer;
+
+/**
+ * We used to have a full index of the resources present in the classpath stored in ClassLoaderState of QuarkusClassLoader.
+ * <p>
+ * However it could lead to some major memory consumption, the index being more than 13 MB on real-life projects.
+ * This was a problem especially since we also keep a list of the resources in each ClassPathElement.
+ * <p>
+ * Going through the whole list of ClassPathElements for each lookup is not a good option so we choose an intermediary approach
+ * of having a lossy index referencing the start of the resource path to reduce the number of ClassPathElements to go through.
+ * This index still returns exact references as we check the ClassPathElements themselves for the presence of the resource.
+ * <p>
+ * In the particular example above, we went down to less than 3 MB for the index, storing ~600 entries instead of 86000+ in the
+ * mapping.
+ * <p>
+ * We try to be clever as to how we build the resource key to reduce the number of misses. It might need further tuning in the
+ * future.
+ */
+public class ClassPathResourceIndex {
```

Comment:
The removal of `ClassLoaderState` and the addition of this class is the whole purpose of this patch. The rest is really to be able to make this happen.

Reply:
It might just be the fact that it's my first day back from PTO, but just by reading the Javadoc, I don't understand how this index is supposed to work. I do understand what problem it tries to solve vs. previous approach, but I don't understand how it does it.

Reply:
It does it by storing a prefix instead of the full class name. A prefix that is deemed to be smart enough to not cause a penalty.
As the `ClassPathElement` contains a list of the resources, we can go through the index to reduce greatly the number of elments to go through and then make sure the resource is there.

Reply:
It's a bit more complex than that as you need to keep full references for some elements (see the sets with local classes, parent first resources...) but that's the general idea of it.

I don't mind adding some javadoc, just tell me if this explanation is good enough for you.